Analysis/src/EventCandidate.cpp

```diff
@@ -0,0 +1,138 @@
+/*! Definition of wrappers for KinFit.
+This file is part of https://github.com/hh-italian-group/h-tautau. */
+
+#include "h-tautau/Analysis/include/EventCandidate.h"
+
+namespace analysis {
+    using LepCandidate = LeptonCandidate<ntuple::TupleLepton>;
+    using LepCollection = std::vector<LepCandidate>;
+    using JetCandidate = Candidate<ntuple::TupleJet>;
+    using JetCollection = std::vector<JetCandidate>;
+    using MET = MissingET<ntuple::TupleMet>;
+
+    EventCandidate::EventCandidate(const ntuple::Event& _event, UncertaintySource _uncertainty_source,
+    UncertaintyScale _scale, analysis::Period _period) : event(_event), uncertainty_source(_uncertainty_source),
+    scale(_scale), period(_period) {}
+
+    LepCollection& EventCandidate::GetLeptons()
+    {
+      if(!lepton_candidates) {
+        CreateLeptons();
+      }
+      return *lepton_candidates;
+    }
+
+    JetCollection& EventCandidate::GetJets()
+    {
+      if(!jet_candidates) {
+          CreateJets();
+      }
+      return *jet_candidates;
+    }
+
+    MET& EventCandidate::GetMET()
+    {
+        if(!met) {
+            if(uncertainty_source == UncertaintySource::TauES)
+                CreateLeptons();
+            else if(jecUncertainties->JetUncertainties_withTotal().count(uncertainty_source))
+                CreateJets();
+            else{
+                tuple_met = std::make_shared<ntuple::TupleMet>(event, MetType::PF);
+                met = std::shared_ptr<MET>(new MET(*tuple_met, tuple_met->cov()));
+            }
+    }
+    return *met;
+    }
+
+    const ntuple::Event& EventCandidate::GetEvent() const
+    {
+        return event;
+    }
+
+    const UncertaintyScale& EventCandidate::GetScale()
+    {
+        return scale;
+    }
+
+    void EventCandidate::CreateLeptons()
+    {
+        double shifted_met_px = 0;
+        double shifted_met_py = 0;
+
+        if(uncertainty_source == UncertaintySource::TauES){
+          tuple_met = std::make_shared<ntuple::TupleMet>(event, MetType::PF);
+          met = std::shared_ptr<MET>(new MET(*tuple_met, tuple_met->cov()));
+        }
+
+
+        lepton_candidates = std::make_shared<LepCollection>();
+        tuple_leptons = std::make_shared<std::vector<ntuple::TupleLepton>>();
+        for(size_t n = 0; n < event.lep_p4.size(); ++n){
+          tuple_leptons->emplace_back(event, n);
+          ntuple::TupleLepton tuple_lepton = ntuple::TupleLepton(event, n);
+          lepton_candidates->push_back(LepCandidate(tuple_lepton,tuple_lepton.iso()));
```

Comment:
it is important to use the same tuple lepton as in tuple_leptons and not the other one.